src/core/algorithms/gfd/gfd_miner.cpp

```diff
@@ -0,0 +1,667 @@
+        } else {
+            vertex_t u = boost::vertex(snd_token.first, pattern);
+            vertex_t v = embedding.at(u);
+            std::map<std::string, std::string> attrs = graph[v].attributes;
+            if (attrs.find(snd_token.second) == attrs.end()) {
+                return false;
+            }
+            snd = attrs.at(snd_token.second);
+        }
+        if (fst != snd) {
+            return false;
+        }
+    }
+    return true;
+}
+
+bool Validate(graph_t const& graph, Gfd const& gfd, Embeddings const& embeddings,
+              std::size_t const sigma) {
+    bool result = false;
+    std::map<vertex_t, std::set<vertex_t>> maps;
+    for (auto& embedding : embeddings) {
+        if (!Satisfied(gfd.GetPremises(), graph, gfd.GetPattern(), embedding)) {
+            continue;
+        }
+        result = true;
+
+        if (!Satisfied(gfd.GetConclusion(), graph, gfd.GetPattern(), embedding)) {
+            return false;
+        }
+
+        BGL_FORALL_VERTICES_T(v, gfd.GetPattern(), graph_t) {
+            if (maps.find(v) != maps.end()) {
+                maps.at(v).insert(embedding.at(v));
+            } else {
+                std::set<vertex_t> vertices = {embedding.at(v)};
+                maps.emplace(v, vertices);
+            }
+        }
+    }
+
+    std::size_t min_size = boost::num_vertices(graph);
+    for (auto& [v, image] : maps) {
+        min_size = image.size() < min_size ? image.size() : min_size;
+    }
+    return result && (min_size >= sigma);
+}
+
+std::vector<Literal> GenerateLiterals(graph_t const& pattern,
+                                      std::map<std::string, Info> const& attrs_info) {
+    std::vector<int> indices(boost::num_vertices(pattern));
+    std::iota(std::begin(indices), std::end(indices), 0);
+    std::vector<std::vector<int>> pairs = GetSubsets<int>(indices, 2);
+
+    std::vector<Literal> result = {};
+    for (auto& index_pair : pairs) {
+        int fst = index_pair.at(0);
+        int snd = index_pair.at(1);
+        std::string fst_label = pattern[boost::vertex(fst, pattern)].attributes.at("label");
+        std::string snd_label = pattern[boost::vertex(snd, pattern)].attributes.at("label");
+
+        for (auto& fst_name : attrs_info.at(fst_label)) {
+            Token fst_token = Token(fst, fst_name.first);
+            for (auto& snd_name : attrs_info.at(snd_label)) {
+                Token snd_token = Token(snd, snd_name.first);
+                result.push_back(Literal(fst_token, snd_token));
+            }
+        }
+    }
+    for (int i = 0; i < static_cast<int>(boost::num_vertices(pattern)); ++i) {
+        std::string label = pattern[boost::vertex(i, pattern)].attributes.at("label");
+        for (auto& name : attrs_info.at(label)) {
+            Token name_token = Token(i, name.first);
+            for (auto& value : name.second) {
+                Token value_token = Token(-1, value);
+                result.push_back(Literal(name_token, value_token));
```

Comment:
`result.reserve(name.second.size() * attrs_info.at(label).size())`

Reply:
Unfortunately, `name.second` can have different number of values, depending on `name.first`. The reserve may not work correctly.